velox/functions/prestosql/json/SimdJsonExtractor.cpp

```diff
@@ -0,0 +1,590 @@
+  std::optional<std::string> extractScalar(const std::string& json);
+  std::optional<std::string> extractFromObject(
+      int path_index,
+      simdjson::dom::object obj);
+  std::optional<std::string> extractFromArray(
+      int path_index,
+      simdjson::dom::array arr);
+  std::optional<std::string> extractOndemand(const std::string& json);
+  std::optional<std::string> extractKeysOndemand(const std::string& json);
+  std::optional<std::string> extractFromObjectOndemand(
+      int path_index,
+      simdjson::ondemand::object obj);
+  std::optional<std::string> extractFromArrayOndemand(
+      int path_index,
+      simdjson::ondemand::array arr);
+  bool isDocBasicType(simdjson::ondemand::document& doc);
+  bool isValueBasicType(
+      simdjson::simdjson_result<simdjson::ondemand::value> rlt);
+
+  // Shouldn't instantiate directly - use getInstance().
+  explicit SimdJsonExtractor(const std::string& path) {
+    if (!tokenize(path)) {
+      VELOX_USER_FAIL("Invalid JSON path: {}", path);
+    }
+  }
+
+ private:
+  bool tokenize(const std::string& path) {
+    if (path.empty()) {
+      return false;
+    }
+    if (!kTokenizer.reset(path)) {
+      return false;
+    }
+
+    while (kTokenizer.hasNext()) {
+      if (auto token = kTokenizer.getNext()) {
+        tokens_.push_back(token.value());
+      } else {
+        tokens_.clear();
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Cache tokenize operations in SimdJsonExtractor across invocations in the
+  // same thread for the same JsonPath.
+  thread_local static std::
+      unordered_map<std::string, std::shared_ptr<SimdJsonExtractor>>
+          kExtractorCache;
+  thread_local static JsonPathTokenizer kTokenizer;
+
+  // Max extractor number in extractor cache
+  static const uint32_t kMaxCacheNum{32};
+
+  std::vector<std::string> tokens_;
+};
+
+thread_local std::unordered_map<std::string, std::shared_ptr<SimdJsonExtractor>>
+    SimdJsonExtractor::kExtractorCache;
+thread_local JsonPathTokenizer SimdJsonExtractor::kTokenizer;
+
+bool SimdJsonExtractor::isDocBasicType(simdjson::ondemand::document& doc) {
+  return (doc.type() == simdjson::ondemand::json_type::number) ||
+      (doc.type() == simdjson::ondemand::json_type::boolean) ||
+      (doc.type() == simdjson::ondemand::json_type::null);
+}
+bool SimdJsonExtractor::isValueBasicType(
+    simdjson::simdjson_result<simdjson::ondemand::value> rlt) {
+  return (rlt.type() == simdjson::ondemand::json_type::number) ||
+      (rlt.type() == simdjson::ondemand::json_type::boolean) ||
+      (rlt.type() == simdjson::ondemand::json_type::null);
+}
+
+std::optional<std::string> SimdJsonExtractor::extractScalar(
+    const std::string& json) {
+  ParserContext ctx(json.data(), json.length());
+  std::string jsonpath = "";
+
+  try {
+    ctx.parseDocument();
+  } catch (simdjson::simdjson_error& e) {
+    // simdjson might throw a conversion error while parsing the input json. In
```

Comment:
When does this happen? Would you give an example? Why is it OK to ignore the error?

Reply:
The exception happens when the json data can't be parsed. Such as json is invalid, like `{"k1":}`.
In presto, jsonExtractScalar() should return `false` in this case, and the output(std::optional<std::string>) should be std::nullopt.

Reply:
Got it. Thank you for clarifying. Perhaps, update the comment to make it a bit clearer.

```
// Return 'null' if JSON input is not valid.
```